### PR TITLE
lxd/device/tpm: Expose tpmrm to the container

### LINKED ISSUE #11138: in containers only tpm device is provided without matching tpmrm
# Required information

 * Distribution: Ubuntu
 * Distribution version: 22.04 
 * The output of "lxc info" or if that fails:
$ lxc info | grep version
api_version: "1.0"
  driver_version: 5.0.1 | 7.1.0
  kernel_version: 5.15.0-52-generic
  os_version: "22.04"
  server_version: "5.8"

# Issue description

Using tpm in the container is hard. With VMs there is kernel provided abrm provided as `/dev/tpmrm0`.

Lxd sets up tpm0 without an access broker / resource manager. Which then leads down the rabbit hole of trying to create one inside the container.

# Steps to reproduce

 1. `lxc profile device add default tpm tpm path=/dev/tpm0`
 2. `lxc launch ubuntu:kinetic`
 3. `lxc exec ls /dev/tpmrm0` not found

# Information to attach

swtpm chardev creates `/dev/tpm1` for example, but on recent enough kernels resource manager device is created as well. Please bind it into the containers as well.

In the kernel rm device number is allocated as follows:
```
tpm.h:#define TPM_NUM_DEVICES           65536
tpm2-space.c:   chip->devs.devt = MKDEV(MAJOR(tpm_devt), chip->dev_num + TPM_NUM_DEVICES);
```

Or possibly we can read/stat check it on the host, or maybe swtpm should do that and print it. not sure.

### PR DESCRIPTION
**Pull request by xnox**
This fixes #11138.

### CODE REVIEW

File: lxd/device/tpm.go [hunk 1]
```diff
@@ -165,12 +167,22 @@ func (d *tpm) startContainer() (*deviceConfig.RunConfig, error) {
 		return nil, fmt.Errorf("Failed to get TPM device information")
 	}
```
Comment (tomponline) [inline lxd/device/tpm.go:169]: What are these numbers? Can we define them as constants? Or comment to explain what the logic is doing here for future readers?

File: lxd/device/tpm.go [hunk 2]
```diff
@@ -38,8 +38,10 @@ func (d *tpm) validateConfig(instConf instance.ConfigReader) error {
 
 	if instConf.Type() == instancetype.Container {
 		rules["path"] = validate.IsNotEmpty
+		rules["pathrm"] = validate.IsNotEmpty
 	} else {
 		rules["path"] = validate.Optional(validate.IsNotEmpty)
+		rules["pathrm"] = validate.Optional(validate.IsNotEmpty)
```
Comment (tomponline) [inline lxd/device/tpm.go:44]: What is "rm" and can it always be derived from path? Do we need a new setting for it?

**Review by tomponline — CHANGES_REQUESTED** [review]
From looking at the linked issue "rm" in this context appears to mean Resource Manager. It would be good to make this clearer in the code by way of comments to make it less opaque for future readers. Maybe something along the lines of describing what a resource manager is and when it is provided by the kernel? Thanks

File: lxd/device/tpm.go [hunk 3]
```diff
@@ -165,12 +167,22 @@ func (d *tpm) startContainer() (*deviceConfig.RunConfig, error) {
...
+		minorRM = 65536 + minor
+	}
+
 	runConf := deviceConfig.RunConfig{}
 
 	err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, uint32(major), uint32(minor), d.config["path"], false, &runConf)
 	if err != nil {
 		return nil, fmt.Errorf("Failed to setup unix device: %w", err)
 	}
+	err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, uint32(major), uint32(minorRM), d.config["pathrm"], false, &runConf)
```
Comment (tomponline) [inline lxd/device/tpm.go:182]: I think this lack of empty line between blocks will fail our static analysis

**Comment by xnox** [comment]
> From looking at the linked issue "rm" in this context appears to mean Resource Manager. It would be good to make this clearer in the code by way of comments to make it less opaque for future readers. Maybe something along the lines of describing what a resource manager is and when it is provided by the kernel? Thanks

yes RM is resource manager. One can either run tpm2-abrmd themselves, or use kernel provided one. These days there are little difference between the two and most use the kernel provided one by default. Without one, the tpm provided inside the lxd container currently is useless and cannot be used beyond the most simple operations. An operation to create a key requires a resource manager, and normally it's just there and automatically used. Given there isn't one exposed in the containers today, imho the tpm2 inside the container is broken. Running tpm2-abrmd inside the container seems to be very hard / broken, given that `/dev/tpm*` inside the container is not owned by the correct user.

Btw, is there a way to specify desired user/group for chrdevices that lxd is mounting inside the cointainers?

In theory, one can derive tpmrm path from tpm, i.e. on linux host they are normally named as a pair of tpmN and tpmrmN. I don't know why lxd configuration for the containers allows to specify an arbitrary path for the first one. And given it can be an arbitrary path today, there is no way to know what RM path should be for it.

We could replace `path:string` with `devnum:uint` such that users specify `device add mytpm tpm devnum=3` which would then create `/dev/tpm3 /dev/tpmrm3` inside the container.

About numbers.... they really seem to be special/random

https://github.com/torvalds/linux/blob/master/drivers/char/tpm/tpm-chip.c#L346
https://github.com/torvalds/linux/blob/master/drivers/char/tpm/tpm2-space.c#L618
https://github.com/torvalds/linux/blob/master/drivers/char/tpm/tpm.h#L34
https://github.com/torvalds/linux/blob/master/drivers/char/tpm/tpm.h#L36

devno 0, tpm device minor TPM_MINOR (244), tpmrm device minor DEVNO + TPM_NUM_DEVICES(65536) = 65536
devno 1, tpm device minor DEVNO (1), tpm device minor DEVNO + TPM_NUM_DEVICES(65536) = 65537
devno 2, tpm device minor DEVNO (2), tpm device minor DEVNO + TPM_NUM_DEVICES(65536) = 65538
...

**Comment by xnox** [comment]
Can Jenkins please be executed on this branch?

**Comment by tomponline** [comment]
jenkins: test this please

**Review by tomponline — APPROVED** [review]
LGTM, thanks!

**Comment by tomponline** [comment]
> > From looking at the linked issue "rm" in this context appears to mean Resource Manager. It would be good to make this clearer in the code by way of comments to make it less opaque for future readers. Maybe something along the lines of describing what a resource manager is and when it is provided by the kernel? Thanks
> 
> yes RM is resource manager. One can either run tpm2-abrmd themselves, or use kernel provided one. These days there are little difference between the two and most use the kernel provided one by default. Without one, the tpm provided inside the lxd container currently is useless and cannot be used beyond the most simple operations. An operation to create a key requires a resource manager, and normally it's just there and automatically used. Given there isn't one exposed in the containers today, imho the tpm2 inside the container is broken. Running tpm2-abrmd inside the container seems to be very hard / broken, given that `/dev/tpm*` inside the container is not owned by the correct user.
> 
> Btw, is there a way to specify desired user/group for chrdevices that lxd is mounting inside the cointainers?
> 
> In theory, one can derive tpmrm path from tpm, i.e. on linux host they are normally named as a pair of tpmN and tpmrmN. I don't know why lxd configuration for the containers allows to specify an arbitrary path for the first one. And given it can be an arbitrary path today, there is no way to know what RM path should be for it.
> 
> We could replace `path:string` with `devnum:uint` such that users specify `device add mytpm tpm devnum=3` which would then create `/dev/tpm3 /dev/tpmrm3` inside the container.
> 
> About numbers.... they really seem to be special/random
> 
> https://github.com/torvalds/linux/blob/master/drivers/char/tpm/tpm-chip.c#L346 https://github.com/torvalds/linux/blob/master/drivers/char/tpm/tpm2-space.c#L618 https://github.com/torvalds/linux/blob/master/drivers/char/tpm/tpm.h#L34 https://github.com/torvalds/linux/blob/master/drivers/char/tpm/tpm.h#L36
> 
> devno 0, tpm device minor TPM_MINOR (244), tpmrm device minor DEVNO + TPM_NUM_DEVICES(65536) = 65536 devno 1, tpm device minor DEVNO (1), tpm device minor DEVNO + TPM_NUM_DEVICES(65536) = 65537 devno 2, tpm device minor DEVNO (2), tpm device minor DEVNO + TPM_NUM_DEVICES(65536) = 65538 ...

Thanks for the explanation :)

**Comment by xnox** [comment]
> > Btw, is there a way to specify desired user/group for chrdevices that lxd is mounting inside the cointainers?
> 
> We normally have a uid and gid settings on unix devices, e.g. https://linuxcontainers.org/lxd/docs/master/reference/devices_unix_char/#devices-unix-char
> 
> So we probably could do with having those settings on the `tpm` device too (only for containers though).

but that's numeric uids only, unfortunately tss user/group are not globally defined, but are dynamically added inside the container. And thus one has to look up names inside the container to make it work. I have now uploaded tpm-udev package update into lunar to address this in the containers https://launchpad.net/ubuntu/+source/tpm-udev/0.6ubuntu1 by adding tpmfiles.d and .path unit to ensure that dynamically added tpms and those that are available at instance start time have correct permissions.

**Comment by tomponline** [comment]
LXD doesn't trust what is inside an instance so we do not perform name to uid resolution using the contents of an instance.

**Comment by tomponline** [comment]
> I have now uploaded tpm-udev package update into lunar to address this in the containers https://launchpad.net/ubuntu/+source/tpm-udev/0.6ubuntu1 by adding tpmfiles.d and .path unit to ensure that dynamically added tpms and those that are available at instance start time have correct permissions.

Thanks!

**Pull request merged**